crates/symbolicator-service/src/download/index.rs

```diff
@@ -0,0 +1,244 @@
+        match result.into_contents() {
+            Ok(segment) => index.append(segment),
+            Err(e) => {
+                tracing::error!(
+                    error = &e as &dyn std::error::Error,
+                    "Failed to download index segment",
+                );
+            }
+        }
+    }
+
+    index.write(file)?;
+
+    Ok(())
+}
+
+impl CacheItemRequest for FetchSymstoreIndexSegment {
+    type Item = SymstoreIndex;
+
+    const VERSIONS: CacheVersions = SYMSTORE_INDEX_VERSIONS;
+
+    fn compute<'a>(&'a self, temp_file: &'a mut NamedTempFile) -> BoxFuture<'a, CacheContents> {
+        let downloader = Arc::clone(&self.downloader);
+        let source = Arc::clone(&self.source);
+
+        Box::pin(download_index_segment(
+            downloader,
+            source,
+            self.segment_id,
+            temp_file.as_file_mut(),
+        ))
+    }
+
+    fn load(&self, data: ByteView<'static>) -> CacheContents<Self::Item> {
+        let _result = data.hint(AccessPattern::Sequential);
+        let index = Self::Item::load(&data)?;
+        Ok(index)
+    }
+
+    fn weight(item: &Self::Item) -> u32 {
+        item.iter().map(|file| file.len() as u32).sum()
+    }
+}
+
+impl CacheItemRequest for FetchSymstoreIndex {
+    type Item = SymstoreIndex;
+
+    const VERSIONS: CacheVersions = SYMSTORE_INDEX_VERSIONS;
+
+    fn compute<'a>(&'a self, temp_file: &'a mut NamedTempFile) -> BoxFuture<'a, CacheContents> {
+        let downloader = Arc::clone(&self.downloader);
+        let source = Arc::clone(&self.source);
+
+        Box::pin(download_full_index(
+            self.segment_cache.clone(),
+            downloader,
+            source,
+            self.scope.clone(),
+            self.last_id,
+            temp_file.as_file_mut(),
+        ))
+    }
+
+    fn load(&self, data: ByteView<'static>) -> CacheContents<Self::Item> {
+        let _result = data.hint(AccessPattern::Sequential);
+        let index = Self::Item::load(&data)?;
+        Ok(index)
+    }
+
+    fn weight(item: &Self::Item) -> u32 {
+        item.iter().map(|file| file.len() as u32).sum()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SymstoreIndexService {
+    cache: Arc<Cacher<FetchSymstoreIndex>>,
+    segment_cache: Arc<Cacher<FetchSymstoreIndexSegment>>,
+    downloader: Arc<DownloadService>,
+    last_id_cache: moka::future::Cache<Url, CacheContents<u32>>,
+}
+
+impl SymstoreIndexService {
+    pub fn new(
+        cache: Cache,
+        shared_cache: SharedCacheRef,
+        downloader: Arc<DownloadService>,
+    ) -> Self {
+        let last_id_cache = moka::future::Cache::builder()
+            .max_capacity(100)
+            .name("last_id")
+            .time_to_live(Duration::from_secs(24 * 60 * 60))
+            .build();
+
+        Self {
+            cache: Arc::new(Cacher::new(cache.clone(), shared_cache.clone())),
+            segment_cache: Arc::new(Cacher::new(cache, shared_cache)),
```

Comment:
We do something here that doesn't happen anywhere else in Symbolicator, namely use the same configuration (including the directory!) for two "different" caches. I actually think this is fairly elegant, it lets us put the individual parts of the index and the entire index in the same cache.

Reply:
Maybe makes sense to put this explanation in the code